Add TI J721e R5 and BeagleBone AI64 R5 initial support

slpp95prashanth opened the pull request:
This PR (non-working port) adds device tree and soc support for TI J721e_r5 and Beaglebone AI64 board support for hello_world application.

TRM for J721e https://www.ti.com/lit/zip/spruil1
File: spruil1c.pdf

BeagleBone AI_64 https://beagleboard.org/ai-64

Signed-off-by: Prashanth S <slpp95prashanth@yahoo.com>

slpp95prashanth committed:
dts: j721e: Add TI J721E R5 device tree

Add initial Device Tree support for the TI J721E SoC series.

TRM for J721e https://www.ti.com/lit/zip/spruil1
File: spruil1c.pdf

Signed-off-by: Prashanth S <slpp95prashanth@yahoo.com>

slpp95prashanth committed:
boards: beaglebone_ai64: Add BeagleBone AI-64

Add initial BeagleBone AI-64 support

BeagleBone AI_64 https://beagleboard.org/ai-64

Signed-off-by: Prashanth S <slpp95prashanth@yahoo.com>

vaishnavachath commented on boards/arm/beaglebone_ai64/bbai_64.dts:
```diff
@@ -0,0 +1,25 @@
+	status = "okay";
+	/* pinctrl-0 = <&uart0_default>;
+	 * pinctrl-names = "default";
+	 */
```
Any reason for skipping this, is it due to missing RAT configurations?

slpp95prashanth replied:
Added pinctrl modules

vaishnavachath commented on dts/arm/ti/j721e.dtsi:
```diff
@@ -0,0 +1,23 @@
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	uart0: uart@2800000 {
```
This UART is the same one being used by Linux, ideally this should not be used in R5 firmware, can you use another UART and update?

slpp95prashanth replied:
Ok, I will update to UART 1

slpp95prashanth replied:
Added Uart 2, as beaglebone ai64 has uart2 pinouts in p8 header.

vaishnavachath commented on soc/arm/ti_k3/j721e/linker.ld:
```diff
@@ -0,0 +1,20 @@
+
+    .log_shared_mem :
+    {
+        *(.log_shared_mem*)
```
Are you using this log buffer? or have plans to use this in future?

slpp95prashanth replied:
In this commits this log buffer is not used. In future it can be used for sending logs to linux.

vaishnavachath replied:
Can we drop the  RAM_CONSOLE section, I don't see why this should be part of platform support especially since it is not being used now.

cfriedt commented on boards/arm/beaglebone_ai64/bbai_64.yaml:
```diff
@@ -0,0 +1,16 @@
+# Copyright (C) 2023 BeagleBoard.org Foundation
```
The files in the board directory need to match the name of the parent directory, from what I remember.

So instead of bbai_64 use beaglebone_ai64 it vice versa

slpp95prashanth replied:
Changing board directory beaglebone_ai64 to bbai_64

cfriedt reviewed (commented):
I thought that filenames in the board directory should match the board name. I think that's a convention.

slpp95prashanth committed:
soc: ti_k3: Add TI J721E SoC R5

Add initial SoC support for the TI J721E SoC series Cortex-R5 core.

TRM for J721e https://www.ti.com/lit/zip/spruil1
File: spruil1c.pdf

Signed-off-by: Prashanth S <slpp95prashanth@yahoo.com>

carlocaione reviewed (changes requested):
I'm quite strongly opinionated about adding support in-tree for a board without an interrupt controller and only with the `hello world` sample running. The feature set for a new SoC / board should be enough to **at least** pass the kernel tests.

carlocaione commented on soc/arm/ti_k3/j721e/linker.ld:
```diff
@@ -0,0 +1,20 @@
+
+SECTIONS
+{
+    SECTION_PROLOGUE(.resource_table,,)
+    {
+        KEEP(*(.resource_table*))
+    } > RAM
```
Why is this RPMSg stuff is needed?

slpp95prashanth replied:
J721e SoC is a heterogeneous has A72 (Linux), cortex R5 (Zephyr). The zephyr image is loaded to cortex r5 through remoteproc from A72 running Linux, so remoteproc requires .resource_table section in the zephyr.elf file.

carlocaione commented on soc/arm/ti_k3/j721e/soc.c:
```diff
@@ -0,0 +1,97 @@
+	return 0;
+}
+
+uint32_t sys_clock_elapsed(void)
+{
+	return 0;
+}
+
+uint32_t sys_clock_cycle_get_32(void)
+{
+	return 0;
+}
+
+uint64_t sys_clock_cycle_get_64(void)
+{
+	return 0;
+}
+
+void z_soc_irq_init(void)
+{
+}
+
+void z_soc_irq_enable(unsigned int irq)
+{
+}
+
+void z_soc_irq_disable(unsigned int irq)
+{
+}
+
+int z_soc_irq_is_enabled(unsigned int irq)
+{
+	return 0;
+}
+
+void z_soc_irq_priority_set(unsigned int irq, unsigned int prio, unsigned int flags)
+{
+}
```
Sorry, this is not acceptable. A full interrupt controller driver is definitely needed before the SoC is considered for inclusion in-tree.

slpp95prashanth replied:
I will add the interrupt controller support as part of this PR.

vaishnavachath replied:
@slpp95prashanth can you also please mark the PR as draft until then?

DhruvaG2000 reviewed (changes requested):
Good job adding support for VIM for R5!
Have added some suggestions mostly for documentation.
I briefly went over the VIM driver and the way you are servicing IRQs Through Vector Interface seems OK, from register definitions POV. I am yet to test and review the exact sequence but if you have followed the method mentioned in the TRM/ MCU SDK as is then it should work fine.

Overall other than the name (bbai_64) used I don't have any major objections.

carlocaione commented on drivers/interrupt_controller/Kconfig.vim:
```diff
@@ -0,0 +1,11 @@
+
+if CPU_CORTEX
+
+config VIM
+	bool
```
Please, check how this is done for the other interrupt controllers. You are missing description, dependencies, etc...

carlocaione commented on drivers/interrupt_controller/intc_vim.c:
```diff
@@ -0,0 +1,140 @@
+{
+	uint32_t num_of_irqs = sys_read32(VIM_INFO_INTERRUPTS_MASK);
+
+	printk("VIM: Number of IRQs = %u\n", num_of_irqs);
```
Please, no `printk` in the code. Use `LOG_*` if necessary.

carlocaione commented on drivers/interrupt_controller/intc_vim.c:
```diff
@@ -0,0 +1,140 @@
+	uint32_t irq_group_num, irq_bit_num, regval;
+
+	if (irq > CONFIG_NUM_IRQS || prio > 15 || flags > 1) {
+		printk("%s: Invalid argument irq = %u prio = %u flags = %u\n",
+		       __func__, irq, prio, flags);
```
ditto

carlocaione commented on drivers/interrupt_controller/intc_vim.c:
```diff
@@ -0,0 +1,140 @@
+
+	regval = sys_read32(VIM_INTR_EN_SET(irq_group_num));
+
+	return !!(regval & (1 << irq_bit_num));
```
Use `BIT()` macro.

carlocaione commented on drivers/interrupt_controller/intc_vim.c:
```diff
@@ -0,0 +1,140 @@
+{
+	uint32_t irq_group_num, irq_bit_num, regval;
+
+	if (irq > CONFIG_NUM_IRQS || prio > 15 || flags > 1) {
```
Use `define` for hardcoded values.

carlocaione commented on include/zephyr/drivers/interrupt_controller/intc_vim.h:
```diff
@@ -0,0 +1,317 @@
+
+/* RAW */
+
+#define VIM_GRP_RAW_STS_MASK                 (0xFFFFFFFFU)
```
here and all over this file please use `BIT_MASK()` when possible.

carlocaione commented on drivers/interrupt_controller/intc_vim.c:
```diff
@@ -0,0 +1,140 @@
+
+#include <zephyr/drivers/interrupt_controller/intc_vim.h>
+
+unsigned int z_soc_irq_get_active(void)
```
You are polluting the `z_soc_*` namespace. Please check how this done for other interrupt controllers.

carlocaione commented on drivers/interrupt_controller/Kconfig.vim:
```diff
@@ -0,0 +1,15 @@
+
+if CPU_CORTEX
+
+config VIM
+	bool "Vectored Interrupt Manager"
+	depends on ARM_CUSTOM_INTERRUPT_CONTROLLER
+	help
+		The Vectored Interrupt Manager provides hardware assistance for prioritizing
+		and controlling the many interrupt sources for ARM Cortex-family processors.
+
```
this should be `depends on DT_HAS_ARM_VIM_ENABLED`. Check how this is done for the other interrupt controllers.

carlocaione commented on drivers/interrupt_controller/intc_vim.c:
```diff
@@ -0,0 +1,150 @@
+{
+	uint32_t num_of_irqs = sys_read32(VIM_INFO_INTERRUPTS_MASK);
+
+	LOG_PRINTK("VIM: Number of IRQs = %u\n", num_of_irqs);
```
why `LOG_PRINTK` and not `LOG_DBG` or something like that?

slpp95prashanth replied:
Changing to LOG_DBG.

carlocaione commented on include/zephyr/drivers/interrupt_controller/intc_vim.h:
```diff
@@ -0,0 +1,328 @@
+#include <zephyr/dt-bindings/interrupt-controller/arm-vim.h>
+#include <zephyr/sys/util_macro.h>
+
+#define VIM_BASE_ADDR       DT_REG_ADDR(DT_INST(0, arm_vim))
+
+#define VIM_MAX_IRQ_PER_GROUP   (32)
+#define VIM_MAX_GROUP_NUM       ((uint32_t)(CONFIG_NUM_IRQS / VIM_MAX_IRQ_PER_GROUP))
+
+#define VIM_GET_IRQ_GROUP_NUM(n)  ((uint32_t)((n) / VIM_MAX_IRQ_PER_GROUP))
+#define VIM_GET_IRQ_BIT_NUM(n)    ((uint32_t)((n) % VIM_MAX_IRQ_PER_GROUP))
```
This looks oddly indented.

slpp95prashanth replied:
Updating indentation.

carlocaione commented on include/zephyr/drivers/interrupt_controller/intc_vim.h:
```diff
@@ -0,0 +1,328 @@
+
+#define VIM_DEDVEC_RESETVAL                  (0x00000000U)
+
+unsigned int z_vim_irq_get_active(void);
+void z_vim_irq_eoi(unsigned int irq);
+void z_vim_irq_init(void);
+void z_vim_irq_priority_set(unsigned int irq, unsigned int prio, uint32_t flags);
+void z_vim_irq_enable(unsigned int irq);
+void z_vim_irq_disable(unsigned int irq);
+int z_vim_irq_is_enabled(unsigned int irq);
+void z_vim_arm_enter_irq(int irq);
```
Please add the doxygen documentation here.

carlocaione commented on soc/arm/ti_k3/j721e/soc.c:
```diff
@@ -0,0 +1,81 @@
+	return 0;
+}
+
+uint32_t sys_clock_cycle_get_32(void)
+{
+	return 0;
+}
+
+uint64_t sys_clock_cycle_get_64(void)
+{
+	return 0;
+}
```
how are you keeping the time?

slpp95prashanth replied:
systick support is not added.

carlocaione replied:
how are you managing delays or the time in general? like `k_msleep()`?

slpp95prashanth replied:
I am working on the timer support.

carlocaione replied:
ok, but please, next time just clarify in the PR message that this is a non-working port otherwise you will end up pissing maintainers off by forcing them to review something not yet complete and working.

slpp95prashanth replied:
Thank You, I will update the PR message.

carlocaione commented:
@slpp95prashanth is this port in the current state able to pass for example the tests in `tests/kernel` for example?

slpp95prashanth commented:
> @slpp95prashanth is this port in the current state able to pass for example the tests in `tests/kernel` for example?

Yes, all the tests kernel are passed/skipped. But 3 testcases failed.
Error Log:
/__w/zephyr/zephyr/include/zephyr/spinlock.h:281:59: error: unused parameter 'k' [-Werror=unused-parameter]
  281 | static ALWAYS_INLINE void z_spin_onexit(k_spinlock_key_t *k)
      |                                         ~~~~~~~~~~~~~~~~~~^
cc1: all warnings being treated as errors

DhruvaG2000 commented:
> > @slpp95prashanth is this port in the current state able to pass for example the tests in `tests/kernel` for example?
> 
> @carlocaione Yes, all the tests kernel are passed/skipped. But 3 testcases failed (https://github.com/zephyrproject-rtos/zephyr/pull/59191/checks?check_run_id=14976309563). Error Log: /__w/zephyr/zephyr/include/zephyr/spinlock.h:281:59: error: unused parameter 'k' [-Werror=unused-parameter] 281 | static ALWAYS_INLINE void z_spin_onexit(k_spinlock_key_t *k) | ~~~~~~~~~~~~~~~~~~^ cc1: all warnings being treated as errors

Is the PR https://github.com/zephyrproject-rtos/zephyr/pull/60375 aimed at fixing the above issue?

carlocaione commented:
> Is the PR #60375 aimed at fixing the above issue?

It is indeed.

carlocaione commented:
@emob-nordic this is for you.

DhruvaG2000 commented on boards/arm/beagle_bbai_64/doc/index.rst:
```diff
@@ -0,0 +1,91 @@
+Note:
+-----
+As the image is loaded through remoteproc, which is running on Linux. To flash
+Linux image on A72, please refer https://beagleboard.org/latest-images
```
I don't think "flash Linux image on A72" conveys the right message.
Rather, I'd move this above steps or as part of steps,

```
Use the latest Linux image from https://beagleboard.org/latest-images to bootup the BBAI-64. This is required for loading zephyr.elf/binary using remoteproc .
```

slpp95prashanth replied:
I will add this above steps.

DhruvaG2000 commented on boards/arm/beagle_bbai_64/doc/index.rst:
```diff
@@ -0,0 +1,91 @@
+.. _beaglebone_ai64:
```
somewhere in your documentation please mention that the uart logs that you have enabled will appear on which port exactly on the BBAI-64?

slpp95prashanth replied:
Adding uart 2 as console information below steps.

vaishnavachath commented on dts/bindings/interrupt-controller/arm,vim.yaml:
```diff
@@ -0,0 +1,20 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: ARMv7-R Vectored Interrupt Manager
+
```
I think this naming scheme in this file introduces confusion, unlike NVIC, GIC .etc , VIM is not a standard ARM interrupt controller, but a TI specific interrupt controller.

vaishnavachath commented on include/zephyr/drivers/interrupt_controller/intc_vim.h:
```diff
@@ -0,0 +1,383 @@
+#define VIM_PID_MINOR_SHIFT                  (0x00000000U)
+#define VIM_PID_MINOR_RESETVAL               (0x00000001U)
+#define VIM_PID_MINOR_MAX                    (BIT_MASK(6))
+
```
Also is this file derived from any of the existing TI SDK sources? looking at these VID/PID/RTL register definitions it looks like that.

slpp95prashanth replied:
I have added all the register description available for VIM module. Took reference from PDK. Should I remove unused register definitions?

vaishnavachath replied:
Thanks for clarifying, I believe the implementation of VIM you have is feature complete, if you think you will need them in future you can keep it, otherwise let's remove unused ones.

vaishnavachath commented on soc/arm/ti_k3/j721e/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,39 @@
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 25000000
+
```
Can you help us understand how you came up with these values?

slpp95prashanth replied:
I took it from TI PDK.

vaishnavachath replied:
Okay thanks for clarifying, this will make sense here when DM timer support is added.

vaishnavachath commented:
@carlocaione @cfriedt Can the interrupt controller addition be a separate PR from platform support for easier reviewing?
Seeing some examples in the past where it has been done like that : https://github.com/zephyrproject-rtos/zephyr/pull/19865 

@slpp95prashanth The commit message for

carlocaione commented:
> @carlocaione @cfriedt Can the interrupt controller addition be a separate PR from platform support for easier reviewing? Seeing some examples in the past where it has been done like that : #19865

No objections from me.

vaishnavachath commented:
Thanks for confirming @carlocaione @cfriedt .

@slpp95prashanth can you please create a new PR for the VIM interrupt controller addition and mention it as a dependency here.

slpp95prashanth commented:
@vaishnavachath I am working on the timer support and want to add in the later commits. Shall we add timer support in the later commits?

vaishnavachath commented:
> @vaishnavachath I am working on the timer support and want to add in the later commits. Shall we add timer support in the later commits?

@slpp95prashanth , sure, do you mean later commits in this PR? If that also can be a separate PR that would be great, in that way the corresponding driver, bindings will be easier to review and experts from the subsytem reviews them.

When you split the changes for intc/other from this PR make sure to follow these points:
* Run twister tests locally with all the combined changes, for example when you create separate pull request for VIM, the twister run may pass for the particular PR, but would not exercise all the code paths since there might be no consumers. This would eventually lead to us sending fix PRs back and forth which is something that we want to avoid.
* Mention the relationship between both PR so that the reviewers are aware.

slpp95prashanth commented:
> > @vaishnavachath I am working on the timer support and want to add in the later commits. Shall we add timer support in the later commits?
> 
> @slpp95prashanth , sure, do you mean later commits in this PR? If that also can be a separate PR that would be great, in that way the corresponding driver, bindings will be easier to review and experts from the subsytem reviews them.
> 
> When you split the changes for intc/other from this PR make sure to follow these points:
> 
> * Run twister tests locally with all the combined changes, for example when you create separate pull request for VIM, the twister run may pass for the particular PR, but would not exercise all the code paths since there might be no consumers. This would eventually lead to us sending fix PRs back and forth which is something that we want to avoid.
> * Mention the relationship between both PR so that the reviewers are aware.

@vaishnavachath Yes as separate PR, Once this PR is merged I will create a separate PR for systick support. With all the modules and multiple PR simultaneously it is quite difficult for me to manage. So as an non working initial support can I have board, soc, device tree and vim support?

vaishnavachath commented:
> @vaishnavachath Yes as separate PR, Once this PR is merged I will create a separate PR for systick support. With all the modules and multiple PR simultaneously it is quite difficult for me to manage. So as an non working initial support can I have board, soc, device tree and vim support?

@slpp95prashanth Sorry, systick is important for the base support, systick support coming in later is not a feasible plan.

slpp95prashanth commented:
This PR depends on #60856 and #61020.

jadonk commented:
From an organizational point of view, I appreciate you starting with "beagle_", thanks!. I think "bbai64" can be grouped together, not "bbai_64". I think someone could easily have an a72 target, so I suggest making the whole directory string "beagle_bbai64_r5" to make it clear. Similarly, I'm planning "beagle_play_cc1352" for the M4 on the CC1352 and "bealge_play_a53" for the A53 cluster in the AM6254.

slpp95prashanth commented:
@jadonk I have updated the board identifier, files and directories to beagle_bbai64_r5.

carlocaione commented:
should this still be in draft?

slpp95prashanth commented:
> should this still be in draft?

Setting to ready for review.

carlocaione commented on dts/arm/ti/j721e_r5.dtsi:
```diff
@@ -0,0 +1,33 @@
+	};
+
+	soc {
+		atcm: memory@0 {
+			reg = <0x00000000 DT_SIZE_K(32)>;
+		};
+
+		btcm: memory@41010000 {
+			reg = <0x41010000 DT_SIZE_K(32)>;
+		};
```
no compatible here?

slpp95prashanth replied:
yes, it is not compatible removing it.

carlocaione commented on soc/arm/ti_k3/j721e/soc.c:
```diff
@@ -0,0 +1,66 @@
+	z_vim_arm_enter_irq(irq);
+}
+
+SYS_INIT(j721e_init, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
```
not needed

carlocaione commented on soc/arm/ti_k3/j721e/soc.c:
```diff
@@ -0,0 +1,60 @@
+
+#include <zephyr/drivers/interrupt_controller/intc_vim.h>
+
+static int j721e_init(void)
+{
+	return 0;
+}
```
remove if not needed

gmarull replied:
+1, drop if empty

slpp95prashanth replied:
Can we pass NULL as a first argument to SYS_INIT()?

gmarull replied:
what's the point of a no-op SYS_INIT...?

slpp95prashanth replied:
Removing SYS_INIT() and j721e_init().

vaishnavachath replied:
@slpp95prashanth , sorry for the late response, but you will need to atleast unlock the CTRL MMR regions during, all the necessary regions might not always be unlocked by bootloader for you.
See https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/ti_k3/am62x_m4/soc.c#L47

carlocaione commented on soc/arm/ti_k3/j721e/soc.h:
```diff
@@ -0,0 +1,10 @@
+/* Copyright (C) 2023 BeagleBoard.org Foundation
+ * Copyright (C) 2023 S Prashanth
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _TI_K3_J721E_R5_SOC_H_
+#define _TI_K3_J721E_R5_SOC_H_
+
+#endif /* _TI_K3_J721E_R5_SOC_H_ */
```
remove if not needed

slpp95prashanth replied:
soc.h is included in include/zephyr/arch/arm/aarch32/cortex_a_r/cmsis.h

gmarull replied:
Cortex-M/A/R require soc.h, even if empty because of how CMSIS is integrated. Ref. https://github.com/zephyrproject-rtos/zephyr/issues/57246

carlocaione commented on dts/arm/ti/j721e_r5.dtsi:
```diff
@@ -0,0 +1,29 @@
+	};
+
+	soc {
+		atcm: memory@0 {
+			reg = <0x00000000 DT_SIZE_K(32)>;
+		};
```
why this doesn't have an `compatible = "mmio-sram";` compatible?

carlocaione reviewed (dismissed):
I'm tired of this.

Did you at least try to compile this?
```
# west build -b beagle_bbai64_r5 samples/hello_world -p

dts/arm/ti/j721e.dtsi:10:10: fatal error: zephyr/dt-bindings/interrupt-controller/ti-vim.h: No such file or directory
   10 | #include <zephyr/dt-bindings/interrupt-controller/ti-vim.h>
```

I honestly feel you are wasting mine and maintainers time here. I'm going to remove myself from this PR.

slpp95prashanth commented:
> I'm tired of this.
> 
> Did you at least try to compile this?
> 
> ```
> # west build -b beagle_bbai64_r5 samples/hello_world -p
> 
> dts/arm/ti/j721e.dtsi:10:10: fatal error: zephyr/dt-bindings/interrupt-controller/ti-vim.h: No such file or directory
>    10 | #include <zephyr/dt-bindings/interrupt-controller/ti-vim.h>
> ```
> 
> I honestly feel you are wasting mine and maintainers time here. I'm going to remove myself from this PR.

Yes I compiled and tested.

I apologize, but this PR depends on #60856 and #61020 which is mentioned in https://github.com/zephyrproject-rtos/zephyr/pull/59191#issuecomment-1665067686.

jadonk commented on boards/arm/beagle_bbai64_r5/Kconfig.board:
```diff
@@ -0,0 +1,9 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_BBAI64_R5
+	bool "Beaglebone AI-64"
```
BeagleBone should be CamelCase. "BeagleBone" not "Beaglebone".

jadonk commented on boards/arm/beagle_bbai64_r5/doc/index.rst:
```diff
@@ -0,0 +1,96 @@
+
+To build hello world application for BeagleBone AI-64:
+
+``west build -p auto -b beagle_bbai64_r5 samples/hello_world``
```
Do you write the firmware file to /lib/firmware? What filename does it get? Do you set the filename through remoteproc temporarily? Will the firmware restart after reboot?

slpp95prashanth replied:
Do you write the firmware file to /lib/firmware?
Yes, I copy the image to lib/firmware

What filename does it get?
zephyr.elf

Do you set the filename through remoteproc temporarily?
Yes, the image name is passed to remoteproc through sysfs.

Will the firmware restart after reboot?
Firmware does not reboot, if Linux is rebooted. But not if power or boot button is pressed to reset.

jadonk commented on boards/arm/beagle_bbai64_r5/doc/index.rst:
```diff
@@ -0,0 +1,96 @@
+------------------------
+The Example shows how to load an image on Cortex R5FSS0_CORE0 on J721e.
+
+| Copy Zephyr image to the /lib/firmware/ directory.
+| ``cp build/zephyr/zephyr.elf /lib/firmware/``
+|
+| Ensure the Core is not running.
+| ``echo stop > /sys/class/remoteproc/remoteproc18/state``
+|
+| Configuring the image name to the remoteproc module.
+| ``echo zephyr.elf > /sys/class/remoteproc/remoteproc18/firmware``
+|
+| Once the image name is configured, send the start command.
+| ``echo start > /sys/class/remoteproc/remoteproc18/state``
```
At one point, we tried to provide /dev/bone aliases for the various processors to make them consistent. `remoteproc18` isn't guaranteed to stay consistent.

gmarull replied:
why not implement a proper runner so you can `west flash`?

jadonk commented on boards/arm/beagle_bbai64_r5/doc/index.rst:
```diff
@@ -0,0 +1,96 @@
+Console
+-------
+The Zephyr on BeagleBone AI-64 J721E Cortex R5 uses UART 2 (Rx p8.22, Tx p8.34)
+as console.
```
Do you have a recommendation for looping it back to the A72?

slpp95prashanth replied:
I think there is a separate GSoC project in ideas page https://elinux.org/BeagleBoard/GSoC/Ideas-2023#remoteproc_tty

DhruvaG2000 commented on boards/arm/beagle_bbai64_r5/doc/index.rst:
```diff
@@ -0,0 +1,96 @@
+
+Overview
+********
+BeagleBone AI-64 is a computational platform powered by TI J721E Soc, which is
```
s/Soc/SoC

DhruvaG2000 commented on boards/arm/beagle_bbai64_r5/doc/index.rst:
```diff
@@ -0,0 +1,96 @@
+| UART      | on-chip    | serial               |
++-----------+------------+----------------------+
+
+Other hardwares features are currently not supported.
```
Add GPIO here once that PR is merged and tested

DhruvaG2000 commented on dts/arm/ti/j721e-pinctrl.dtsi:
```diff
@@ -0,0 +1,25 @@
+
+&pinctrl {
+	uart0_tx_default: uart0_tx_default {
+		pinmux = <0x1e8 (PIN_OUTPUT | 0)>;
```
Adding a comment as to which pin this corresponds to would also help
```
pinmux = <0x1e8 (PIN_OUTPUT | 0)>; /* add comment here P8.xx header and mux mode */

DhruvaG2000 commented on dts/arm/ti/j721e-pinctrl.dtsi:
```diff
@@ -0,0 +1,25 @@
+	};
+
+	uart2_tx_default: uart2_tx_default {
+		pinmux = <0x1c (PIN_OUTPUT | 14)>;
```
Why is it 14? What pin is 0x1c? Add comments please.

DhruvaG2000 replied:
Or else make it as some #defines, as long as a person reading the code doesn't have to dig around TRMs / datasheets to just understand what 14 and 0 correspond to

DhruvaG2000 commented on dts/arm/ti/j721e-pinctrl.dtsi:
```diff
@@ -0,0 +1,25 @@
+		pinmux = <0x1e8 (PIN_OUTPUT | 0)>;
+	};
+
+	uart0_rx_default: uart0_rx_default {
```
I don't see ``uart0_*_default`` being used anywhere else, what is the purpose of adding it then?

slpp95prashanth replied:
removing it.

gmarull commented on boards/arm/beagle_bbai64_r5/beagle_bbai64_r5_defconfig:
```diff
@@ -0,0 +1,27 @@
+# enable uart
+CONFIG_SERIAL=y
+
+# enable console
+CONFIG_CONSOLE=y
+CONFIG_UART_CONSOLE=y
+
+CONFIG_BUILD_OUTPUT_UF2=n
+CONFIG_BUILD_OUTPUT_HEX=y
+
+CONFIG_XIP=n
+CONFIG_FLASH_SIZE=0
+CONFIG_FLASH_BASE_ADDRESS=0
+
+CONFIG_BOOTLOADER_SRAM_SIZE=0
+
+CONFIG_PRINTK=y
+
+CONFIG_OPENAMP_RSC_TABLE=y
```
please review this list, it should be minimal per porting guidelines

gmarull commented on boards/arm/beagle_bbai64_r5/doc/index.rst:
```diff
@@ -0,0 +1,96 @@
+
+Other hardwares features are currently not supported.
+
+The default configuration can be found in the defconfig file:
+
+``boards/arm/beagle_bbai64_r5/beagle_bbai64_r5_defconfig``
```
redundant

gmarull commented on boards/arm/beagle_bbai64_r5/doc/index.rst:
```diff
@@ -0,0 +1,96 @@
+
+``boards/arm/beagle_bbai64_r5/beagle_bbai64_r5_defconfig``
+
+Flashing
+********
+The Zephyr image can be flashed to J721E Cortex R5 through remoteproc from
+Linux (Running on A72).
+
+To build hello world application for BeagleBone AI-64:
+
+``west build -p auto -b beagle_bbai64_r5 samples/hello_world``
```
please avoid trivial documentation, this is generic to any board

gmarull commented on dts/arm/ti/j721e.dtsi:
```diff
@@ -0,0 +1,41 @@
+		interrupts = <0 160 IRQ_TYPE_LEVEL IRQ_DEFAULT_PRIORITY>;
+		interrupt-parent = <&vim>;
+		reg-shift = <2>;
+		status = "okay";
+	};
+
+	systick_timer: timer@24c0000 {
+		compatible = "ti,am654-dmtimer";
+		reg = <0x24c0000 0x70>;
+		reg-shift = <2>;
+		interrupts = <0 168 IRQ_TYPE_LEVEL IRQ_DEFAULT_PRIORITY>;
+		interrupt-parent = <&vim>;
+		status = "okay";
```
please keep nodes disabled in soc dts files

slpp95prashanth replied:
This PR depends on #60856 (VIM Interrupt controller) and #61020 (Systick Timer) PRs.

gmarull replied:
then please, fix the issue there.

slpp95prashanth replied:
Can I make it disabled and once the systick timer and vim interrupt controller is merged, Can I create a separate commit for enabling these modules?

gmarull replied:
it's board responsability to enable these nodes (or app in some cases)

gmarull commented on soc/arm/ti_k3/Kconfig:
```diff
@@ -9,7 +9,6 @@ config SOC_FAMILY_TI_K3
 	bool
 
 if SOC_FAMILY_TI_K3
-
```
?

slpp95prashanth replied:
@KarthikL1729 can you please update this?

gmarull commented on soc/arm/ti_k3/j721e/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,39 @@
+	default UART_NS16550_VARIANT_NS16750
+endchoice
+
+config PINCTRL
+	default y
```
-1, please make sure drivers select this

slpp95prashanth replied:
@gmarull should I remove this from here, and add "select PINCTRL" in drivers.
